[WIP, RFC] doc/memos: Added RDM on high level timer API requirements and common features

maribu opened the pull request:
### Contribution description

Early draft for an RDM on a high level timer API.

### Testing procedure

n/a

### Issues/PRs references

None

maribu commented:
I think I now requested reviews from every maintainer that contributed to either xtimer or ztimer and every reviewer of ztimer.

miri64 commented:
If it is a draft, please keep to the [naming scheme for drafts](https://github.com/RIOT-OS/RIOT/blob/master/doc/memos/rdm0000.md#3-memo-publishing-and-maintenance-process):

> Initially, a proposed RDM is identified by its temporary file name of shape rdm-draft--.md (for example: rdm-draft-baccelli-rdm-format.md) instead of a number assigned at a later stage. At this stage, the RDM's header indactes status "draft".

maribu commented:
> If it is a draft, please keep to the [naming scheme for drafts](https://github.com/RIOT-OS/RIOT/blob/master/doc/memos/rdm0000.md#3-memo-publishing-and-maintenance-process):

Done

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+Actual timer hardware however is difficult to use. One main reason is that
+timer hardware can differ significantly starting with the clock frequency,
+the timer width (8 bit, 16 bit, 24 bit, and 32 bit are all common), and their
+time representation. Another reason is that hardware timers only allow setting
```
multiple capture compare registers?

kaspar030 replied:
maybe change to "allow setting a (very) small number of alarm callbacks"?

maribu replied:
Done

maribu replied:
I updated the formulation to the given suggestion.

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+- The implementation must only disable IRQs for short periods of time
+- The implementation must cause as little IRQs as possible and its ISRs must
+  be as short as possible
+- The implementation should provide means to execute callbacks in thread
```
this is IMO out of scope

kaspar030 replied:
... or should maybe go to the utility functionality...

maribu replied:
> or should maybe go to the utility functionality

Yep. It is already in the feature requirements below utilities. To me, this is an important thing to improve real time capabilities in a convenient way. (Obviously, one could just write their own handlers to call `event_post()`; but I think having a utility function for that can be much easier to use.)

I'm not sure if it makes sense to have this written twice. To me this is both a feature requirement and a realtime requirement; so I just added it twice.

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time
```
I'm pretty sure that at least for list-based timers, e.g. using a mutex for the list operations is definitely possible, but harder (more complex) than just disabling IRQs. It might also imply a tradeoff not only in code simplicity but also runtime / code size. So, I'm not happy with the "must" here.

maribu replied:
The "short periods of time" is deliberately a bit vague to give the implementation some freedom in this regard. But there is clearly an advantage in having IRQs only disabled for a bounded time for real time tasks. This bound could also be an `O(n)` with `n` being the worst cast number of software timers used concurrently; any specific use case will have a specific fixed `n` there, so that the worst case time of a timer implementation for that specific use case can be calculated and real timer properties can be verified.

maribu replied:
How about this? (See latest fixup)

kaspar030 replied:
Better. :)

I think `O(n)` hardly qualifies as "real-time", but as we're are talking "requirements", "real-time" would certainly be nice-to-have, but the trade-offs involved might make it infeasible as default implementation, if we'd go stricter than `O(n)`.

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time
+- The implementation must cause as little IRQs as possible and its ISRs must
```
Here I'm also not sure about the "must", and "as little as possible".

How little is that? One ISR per callback? Less than that? Why not use a counting thread, triggering no ISR? Or is it acceptable to set intermediate timers to extend small width timers?
How bad is an extra ISR every 32.7ms, or 24.5minutes?

Again, there are tradeoffs involved.

maribu replied:
How about the new wording?

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+
+## Precision Requirements
+
+- Timer callbacks should never fire prematurely (+- 1 tick of the underlying
```
Important point. Is "-1" acceptable?

maribu replied:
With that formulation, clearly yes. I think the effort needed to prevent this could be quite significant. So if no one cares about it triggering early by 1 tick enough to speak up, we should just allow this :-)

kaspar030 replied:
> I think the effort needed to prevent this could be quite significant.

`set(val) { val += 1; ... }`, or `run_callback() { while(now() < target_time) {}; ...` can achieve this.

We had a discussion some time ago, where people convinced me that "at least as" semantics are the only sane ones. Don't care about one tick? Set the timeout one less. (Having written that, I realize it also works the other way. :) )

The impact of this is very different depending on us, ms, or even second ticks are used.

I'm OK with keeping `-1` for now.

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+
+- Timer callbacks should never fire prematurely (+- 1 tick of the underlying
+  timer)
+- The accuracy of execution delay should be measured with single digit micro
```
You mean, "timer_delay(us)" should have single digit us precision?

maribu replied:
Yes, but with [accuracy](https://en.wikipedia.org/wiki/Accuracy_and_precision), not precision. I added some common sense to the formulation to avoid that someone running an 8 bit platform under high load with an 1 kHz timer is expecting to achieve <10µs accuracy for short delay.

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+  timer)
+- The accuracy of execution delay should be measured with single digit micro
+  seconds for short delays
+- Timer callbacks should fire as early as possible after the timer expired
```
Mmmmh, "as early as possible" would mean, forbid any house keeping from timer ISR entry to callback execution (means: set timer callback as ISR). That's not a good requirement.
Also, inaccuracies or overhead can be compensated.

maribu replied:
I updated the formulation, please check again

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+- Optimized for power management
+    - The timer API should only cause the system to wake up when strictly needed
+- Efficient use of CPU cycles
+
```
Apart from the "wakeup only when strictly needed", (which is vague: is it ok to wakeup for intermediate timers, e.g., on a 16bit 32768hz RTT?), I think "efficient use of" is a given. The tradeoffs or bounds are to be defined.

maribu replied:
I changed "wakeup only when strictly needed" to "minimize system wake ups". With this formulation trade-offs should be possible, while still having reduced power consumption an explicit goal.

kaspar030 replied:
perfect

kaspar030 commented:
@maribu Thanks for getting this rolling!

I've commented on what came to mind while reading.

emmanuelsearch commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+    - [TOSThreads Tutorial](http://tinyos.stanford.edu/tinyos-wiki/index.php/TOSThreads_Tutorial)
+- Mynewt:
+    - [OS Time API reference](https://mynewt.apache.org/latest/os/core_os/time/os_time.html)
+
```
@maribu @kaspar030 Makes sense to add RIOT and xtimer here, does it not?

maribu replied:
I added xtimer. I guess it also makes sense to add power management as metric; but that will take some time to research for all the other OSs

emmanuelsearch commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+| Context of Timer Callback | Thread    | ?         | ?         | Both? (3) | ?         | n.a.      |
+| System Time               | ✓         | ✓         | ✓         | ✓         | ✓         | ✓         |
+| Delays                    | ✓         | ● (1)     | ✓         | ● (4)     | ✓ (5)     | ✓         |
+
```
Same comments as above. (Adding RIOT's current xtimer here as "background") would make sense I think

maribu replied:
Done

emmanuelsearch commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,201 @@
+
+## Contact
+
+TODO: Do we need this?
```
@maribu Yes we do! The goal is to give a handle to people who might have about questions about the doc later down the line...
(similar to an IETF draft or RFC lists author's email/postal address at the end of the doc).

maribu replied:
How about the devel mailing list to reduce the bus factor here?

emmanuelsearch replied:
What mentioning both 
- main author(s) contact(s) and 
- a default: for that would maintainers@ be better than devel@ ?

Let's try to both reduce the bus factor *and* to increase named responsiveness in most cases ;)

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,210 @@
+- The accuracy of execution delay should be measured with single digit micro
+  seconds for short delays in absence of other load on the system
+    - It is acceptable if platform specific tweaking is needed to achieve this
+    - It is acceptable if minimum hardware requirements are needed to achieve
```
This I like ;) It is a nice blank check excuse for unmet performance.

emmanuelsearch commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,216 @@
+
+## Performance Requirements
+
+- Efficient use of RAM/ROM
```
@maribu @kaspar030 Here, and wherever else applicable in this section, what about giving *some* ballpark figures that we aim for, at the minimum?
For that, a (low) hanging fruit could be to use the output of some xtimer benchmarks.
In my mind, such figures would not aim to be "exhaustive", but rather to be indicative.
I think it could be helpful. What do you think?

kaspar030 replied:
#12942 might provide some, but they're very dependent on both platform and configuration...

emmanuelsearch replied:
> very dependent on both platform and configuration...

yup. That's why I was suggesting we do not claim these numbers capture *exhaustively* the requirements, but rather are put here as *indicators*, to give a rough idea?

maribu commented:
I added some more requirements regarding the system timer. I pesonaly have some use cases where I need "portable time stamps", so that I can transfer them over network and both sender and receiver interpret that time stamp in the same way. In practice, UNIX timestamps are most convenient due to their widespread use and are the obvious choice here.

Also: IMO time stamps should be 64 bit. (Unlike the rest of the API, where I consider 64 bit as a waste of resources.) With timestamps, I really don't want to care about overflows...

maribu commented:
> I pesonaly have some use cases where I need "portable time stamps"

I think it didn't become clear what the argument is. The point is, if my personal non-rocket-science projects needs this, this likely is a rather common use case.

jia200x commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,240 @@
+    - Long term delays must play well with power management
+    - Highly precise short term delays (e.g. for bit banging)
+- System time
+    - APIs to get the current system time in reasonable, hardware independent
```
As discussed offline with @kaspar030, a function for setting an absolute timeout time is needed for most time sensitive MAC layers (LoRaWAN, TSCH).

Something like:
```
high_level_timer_set_absolute(&timer_struct, timestamp);
```

Usually the slot times for these protocols are known before hand.

maribu replied:
Done

jia200x commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,240 @@
+        - Note: The rest of the API can be 32 bit, only for time stamps 64 bit
+          support is mandatory
+    - Optional feature to auto-initialize system time from RTC/RTT during boot
+    - APIs to allow implementing network time synchronization
```
+1 here

With LoRaWAN there are some commands to get the absolute timestamp from the network server (with < 10ms of accuracy). Implementing such an API for LoRaWAN would be straightforward.

jia200x commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,240 @@
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+
```
Are time conversion functions in the scope of this RDM? E.g GPS <-> Unix timestamp conversion?

maribu replied:
:+1:

IMO everything related to time can be added to the utilities parts of the timer API, if there is a use case for it. Most of those function will likely only be loosely based on the core timer API, so that they could be retrofitted as follow-up PR to pretty much every core timer API. Thus, it might not be a good criteria for deciding which timer API to choose. But as a road map of features we would like to have it is IMO certainly worth to list this. (Also the timeout on receiving a message or locking a mutex falls into that category.)

maribu replied:
Done

benpicco commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,243 @@
+
+- Efficient use of RAM/ROM
+- Optimized for power management
+    - The timer implementation should minimize system wake ups
```
Most MCUs have two kind of timers

 - coarse, low-power timers
 - high precision timers with higher power requirement

Now a timer API could combine both to always set the longest possible sleep with the low power timer and use the high precision timer for the rest.

e.g. if you have a low power 32kHz timer and a fast 1 MHz timer and want to sleep for 325 ticks and then use the high precision timer for the remaining 82 ticks.

How ideally you could also specify a minimum and a maximum wake up time, so wake-up events can be bundled to reduce the amount of total wake-ups.

For precise timings you'd have `min == max`

maribu replied:
> e.g. if you have a low power 32kHz timer and a fast 1 MHz timer and want to sleep for 325 ticks and then use the high precision timer for the remaining 82 ticks.

I'm not sure if there really is a use case for that. E.g. when you want to perform 3 measurements with a delay of exactly 10 ms and additionally want to repeat those three measurements every 30 minutes, this would be a good example for using both timers. But to me it makes more sense to have one 30 minutes timer, and then a 10 ms timer that is used for the second and third measurement of each row.

If you would transparently use both low frequency and high frequency timers (e.g. for a sleep of 30 minutes and 1337 µs, transparently set one 30 minute low frequency timer and afterwards a 1337 µs high frequency timer), there would be some issues: 

1. The overhead of setting the high frequency timer would reduce precision
2. Low and high frequency domain might be driven from different crystals, so there might be a clock drift between those two. (In fact, having a separate 32.678 kHz crystal for the low frequency timer is quite common.)
3. The chance to start a timer right on the clock tick is pretty low. In the worst case are off by just bellow clock tick. For a 1 MHz timer that would be up to 1 µs precision loss. On a 32.678 kHz timer that would be ~30 µs precision loss.

IMO mixing clocks should not be done transparently, as this might (depending on the availability of low frequency clocks) reduce the accuracy of timers in unexpected ways. Having this explicitly makes sure users are aware of power consumption, accuracy and precision of the used clock.

> How ideally you could also specify a minimum and a maximum wake up time, so wake-up events can be bundled to reduce the amount of total wake-ups.

Sounds like this adds quite some complexity. I'm not sure if it is better to let the application developers optimize their use of timers. IMO core infrastructure like software timers should come with low ROM / RAM requirements and robust implementations, as one usually cannot opt out of using them.

kaspar030 replied:
Yeah, this is something that seems useful in theory. In practice, in a non-idle system, it is difficult enough to get a 1MHz timer actually have better than the 30.5us accuracy that a 32kHz timer promises. See https://github.com/RIOT-OS/RIOT/pull/14931. We tried debugging this, uncorrected, the (quite fast) nrf52840 has 30us offset on a simple ztimer_sleep(). There's a context switch, expensive timer reads, and ztimer's overhead. The time is spread more or less evenly across them. Yeah, that can be mostly corrected, but this is the simplest use-case.

A simple combination of timers won't gain anything, as setting a low-freq timer already quantizes on that timer's frequency, so the result is off by half that timer's period length, on average, anyhow. So using a HF timer "for the last microseconds", but being based on a, from the HF timer's perspective, arbitrary time in between a LF period, is useless.
There are ways to double sync, but that's expensive...

IMO, with all these theoretical timing needs, people need to realize that in most cases, 1KHz or 32786KHz is mightily sufficient. And those are easy to do with RTTs.

kaspar030 replied:
Also, as said in previous discussions, a sane abstraction of timers that use only a single backend (e.g., don't do these complex combinations of timers) are a nice base to put the complex stuff __on top__. At least for prototyping.

kaspar030 replied:
> But the user shouldn't have to worry about which timer to use, that should be selected automatically based on the wait time.

The user needs to choose the precision (the time base). There cannot be an API where the user says "sleep 100". It'll always be "sleep 100ms" or "sleep 100us".

IMO it is OK to simplify our lives (and reduce complexity) by asking the user to always choose the time base that is just enough to express the needs. E.g., if the user asks for "sleep 100ms", it is acceptable to expect them to call "sleep_ms(100)", and not do "sleep_us(100000)".

MichelRottleuthner replied:
Adding another opinion: the arguments @benpicco raised and the underlying ideas should (and can) not be dismissed like that.

> 1. The overhead of setting the high frequency timer would reduce precision

This overhead should be well defined. If this overhead is not tolerable it is also not tolerable for using just the HF timer.

> 2 Low and high frequency domain might be driven from different crystals, so there might be a clock drift between those two. (In fact, having a separate 32.678 kHz crystal for the low frequency timer is quite common.)

True. But not relevant for deciding if splitting across timers (domains) makes sense:
1) You'd know how the timeout is split between both domains, thus drift boundaries of both can be combined to give a respective boundary for the compound.
2) In fact, using the more accurate LF crystal for the significant (long) part of the timeout *improves* the overall drift (versus perfect reference) compared to only using a worse accuracy HF clock.
3) Clock drifts can be determined and compensated or calibrated

This: ..
> The chance to start a timer right on the clock tick is pretty low. In the worst case are off by just bellow clock tick. For a 1 MHz timer that would be up to 1 µs precision loss. On a 32.678 kHz timer that would be ~30 µs precision loss.

.. and this ..

> A simple combination of timers won't gain anything, as setting a low-freq timer already quantizes on that timer's frequency, so the result is off by half that timer's period length, on average, anyhow. So using a HF timer "for the last microseconds", but being based on a, from the HF timer's perspective, arbitrary time in between a LF period, is useless.
There are ways to double sync, but that's expensive...

... seem to be only related to "thinking inside of the box" ;)
Your quantization problem is only there because you assume there is no way to relate the absolute time references of the LF timer to the HF timer. If you know when exactly a timeout of n ticks on the LF timer will fire (i.e. where in the period it is) you can adjust the remaining ticks for the HF timer accordingly.

> IMO, with all these theoretical timing needs, people need to realize that in most cases, 1KHz or 32786KHz is mightily sufficient. And those are easy to do with RTTs.

Agree! Only: "most cases" != "all cases other developers may rightfully care about"
We should not decide what makes sense for others. If something is not possible, well it's not possible. But if it **is** possible and reasonable we should not rule out or restrict it because we have different priorities.

> (..) nice base to put the complex stuff on top

Fully agree as long as the base doesn't limit us to still do the possible & reasonable things others deem more important than the things we consider as "most cases".

> (..) it is acceptable to expect them to call "sleep_ms(100)", and not do "sleep_us(100000)".

Kind of. Still, regarding power consumption one might argue that it would be easier to write *save as much energy as possible*-applications in a platform independent way if the abstraction does whatever is best on the current platform. The restrictions that made you decide xx ms is the best threshold to always map to timer A instead of timer B may not be the same on another platform and the desired timeout may be variable. I thought these kind of considerations are of interest to go for "the RIOT way": make it easy for the developer AND provide flexibility if you want to adapt things more towards something you explicitly need. Maybe my interpretation is off here?

Sometimes writing can be interpreted in many ways, so:  all just my opinion, I don't have negative feelings towards any of you, I'm open minded, we probably all want the best solution :v: 
Are you guys available and interested in some joint discussion of these things during the summit? I don't think we need a separate breakout session for this but maybe chatting about this during a coffe-break or something helps understanding the different perspectives better? What do you think?

kaspar030 replied:
I think this is re-hashing the mailing list discussions from a year ago :)

I'd be happy to see a prototype of what is proposed here.
I agree that for "sleep_ms(100)" vs "sleep_us(100000)", there are ways to maximize power efficiency. A simple `if us > 1000: sleep_ms(us/1000), sleep_us(rest)` might do it for many cases.

Otherwise, I see split between domains, clock drift boundaries, high precision LF, mentioned. Implement! :)

> ... seem to be only related to "thinking inside of the box" ;)
> If you know when exactly a timeout of n ticks on the LF timer will fire (i.e. where in the period it is) you can adjust the remaining ticks for the HF timer accordingly.

Well, how do you do that without explicit synchronization? Assuming LF is being used as it is active during sleep, and HF is not active during sleep, they need to be synchronized *at least* on every transition between sleep and wake. I call that double sync.

I call that expensive because on some platforms (e.g., sam0), just reading the RTT "now" register needs to synchronize the clocks, in some cases taking __a whole period__ (30us). __sometimes__, depending on what else is going on in the system. That dwarfs a lot of clock drift for reasonable time scales.

> In fact, using the more accurate LF crystal for the significant (long) part of the timeout improves the overall drift (versus perfect reference) compared to only using a worse accuracy HF clock.

Yup. I'm strongly in favor of not making things complex, and teach users to use ms or 32kHz, unless there's a reeeeally good reason. Accurate ms or 32kHz timing goes very far. :)

kaspar030 replied:
> Are you guys available and interested in some joint discussion of these things during the summit? I don't think we need a separate breakout session for this but maybe chatting about this during a coffe-break or something helps understanding the different perspectives better? What do you think?

Sure!

MichelRottleuthner replied:
> I think this is re-hashing the mailing list discussions from a year ago :)

I know, and to be honest I don't want to bore people with lengthy discussions again. There is no point in these lengthy "abstract" discussions if not more of the developers jump in.  Yet, the arguments are not less valid than before. It appears to me that we all hold single pieces of the big picture without an explicit agreement on how the big picture looks like. This document will eventually help with that I think.

> Otherwise, I see split between domains, clock drift boundaries, high precision LF, mentioned. Implement! :)

On it (at least for some of these problems) just from a slightly different direction.. I'll explain during the summit.

> Well, how do you do that without explicit synchronization? (..)

Not. At least not without any synchronization. Of course they need to be synchronized.
But why double sync? And who says that you need to do this explicitly? We may very well use an "opportunistic" approach where you sync based on events that happen anyway. I.e., you don't care how long reading 'now' takes for the slow timer if you read the fast timer close enough to a known event of the slow timer... I don't say this solves all cases everywhere. I'm just trying to think towards a solution instead of searching for reasons against it.

> I call that expensive

Yes reading LF timer can indeed be expensive. Apart from reading not being necessarily required, some would argue transitioning to a low power mode for arbitrarily small periods is "expensive" because of the delay when leaving lpm. Again, it all depends on the perspective ;)

> I'm strongly in favor of not making things complex, and teach users to use ms (..)

Fair enough. I get your point.

MichelRottleuthner replied:
BTW: why do I get these check-labels mail-notification after adding a comment? 
> [RIOT-OS/RIOT] Run failed: check-labels - timer-rdm (f5fae3e)

maribu replied:
In the end this will have to be a trade-off between:

1. Overhead in terms of CPU cycles
2. RAM and ROM consumption
3. Complexity and maintainability of the implementation
4. Give the developer using the API a solid understanding and full control about accuracy, resolution, and impact on power consumption for a timer
5. Simplify API so that `sleep_ms(1)` is just an alias for `sleep_us(1000)`
6. Allow optimization of power consumption, accuracy, etc. without replacing code. (E.g. via `USEMODULE += ztimer_policy_powersafe` sleeps could be moved to the low frequency clock where possible, etc. And one would not have to replace `ztimer_sleep(ZTIMER_USEC, x)` by `ztimer_sleep(ZTIMER_MSEC, (x + 500) / 1000)`)

We just won't be able to get those all.

IMO we should first assemble a list design goals and vote on them to get some agreement of a prioritization of them. If overhead, memory, and maintainability get a higher priority than user friendliness and timer policy support, an explicit clock selection via the API makes more sense. Otherwise, adding heuristics and policies for clock selection makes more sense. (This might still be technically better implemented on top of a "low-level high-level timer API", but that would be an implementation detail.)

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,243 @@
+The authors of this memo can be contacted via email at ...
+TODO: Add emails of authors prior merging
+
+Additionally, the RIOT developer community be reached via email at
```
```suggestion
Additionally, the RIOT developer community can be reached via email at
```

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,243 @@
+| Periodic Software Timers  | ✓             | ✓             | ✓         | ● (2)         | ✓         | ✗         | ● (6)     | ● (6)             |
+| Context of Timer Callback | Thread        | ?             | ?         | Both? (3)     | ?         | n.a.      | Interrupt | Interrupt         |
+| System Time               | ✓             | ✓             | ✓         | ✓             | ✓         | ✓         | ✓         | ● (8)             |
+| System Time Width         | `TickType_t`  | 32bit         | 64 bit    |`clock_time_t` | ?         | 96 bit (7)| 64        | 32                |
```
```suggestion
| System Time Width         | `TickType_t`  | 32bit         | 64 bit    |`clock_time_t` | ?         | 96 bit (7)| 64        | 64 (10)             |
```

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,243 @@
+9. "Portable System Time" means that two nodes can exchange time stamps over
+   network and they still refer to the same point in time, assuming the clocks
+   have been synchronized.
+
```
```suggestion
  10. ztimer optionally keep's each clock's time in 64bit (module ztimer_now64)

```

kaspar030 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,244 @@
+9. "Portable System Time" means that two nodes can exchange time stamps over
+   network and they still refer to the same point in time, assuming the clocks
+   have been synchronized.
+  10. ztimer optionally keep's each clock's time in 64bit (module ztimer_now64)
```
```suggestion
10. ztimer optionally keeps each clock's time in 64bit (module ztimer_now64)
```

kaspar030 commented:
IMO this is mostly good to go.

I suggest:

- Let's cut out the design decisions section. It can be part of another RDM proposing a specific implementation, or part of the module level documentation we'll end up with.

- @MichelRottleuthner, maybe add a word or two on what you feel is missing in this RDM, but is not dealt with in the low-level timer RDM.
  Things like combining timers for either reduced latency or automatic choice of a clock depending on sleep duration comes to mind. (If you add something along those lines, I'd prefer those to be marked as "nice to have", as it is not clear whether that can be done efficiently, or what the tradeoffs would be.)
I think we all have agreement on the requirements as listed so far.

@maribu originally it's your PR, do you feel like going over it for last polishing text-wise?

tcschmidt commented:
@kaspar030 if so, then this document should change the title. 
Currently it presents an overview about existing timer APIs in IoT OSs and some requirements, which mainly are neither sharp nor well justified. 
There is no word about concepts and design of a " high level timer API".

ngandrass commented:
Hey everyone,

I recently opened a PR contributing a RDM on a low-level timer-API (#15046) and referenced this PR in there. IMHO it might be useful to cross-reference both RDMs so that one can easily dive deeper into the whole timer topic.

Oh and not to forget: Thank you for the good work that already went into this PR here and please feel free to also leave your feedback on #15046 :)

maribu commented:
> I suggest:
> 
> * Let's cut out the design decisions section. It can be part of another RDM proposing a specific implementation, or part of the module level documentation we'll end up with.

Agreed. On a second thought, I think it might not be the brightest idea to document design decisions via an RDM. While requirements are rather slowly changing, design decisions and the architecture of the implementation are more short lived. Having an up to date Doxygen documentation is clearly more valuable, than an outdated RDM.

> [...] some requirements, which mainly are neither sharp [...]

I assume by not sharp you are referring to the fact that the requirements are qualitative, and not quantitative. This is intentionally. Quantifying a requirement seems to not be worth the effort:

1. It is a lot of effort, as system load, hardware capabilities, and use case have a bigger impact in many cases then the actual implementation.
    - Providing quantitative requirements would require an in-depth specification of the context. And for reasonable results, multiple data points for a reasonable subset of possible combinations of system load, hardware and use case would need to be provided
2. The matter is pressing. RIOT is currently hardly usable for low power scenarios. The two non-academic low power use cases presented in the RIOT summit either disabled the default timer API of RIOT and used `ztimer` instead, or the high level timer API would completely avoided during the sleeps.
    - Rather than having RIOT users resort to fight against the default timer API `xtimer` to meet their low power requirements, we should provide reasonable solutions for those use cases.
    - Low power scenarios are neither obscure and uncommon, nor super recent. There was a big discussion about this at the previous RIOT summit (2019) and it was quite pressing still then. It is now finally time to focus on getting results, rather than quantifying every detail of the requirements
3. Quantitative requirements have zero benefit for deciding for or against a specific proposal
    - We end up picking the best option out of all available.
    - Only how one solution compares relative to each other is really important. This comparison can easily be quantitative by providing benchmark results. And those benchmarks will be of value even after the decision, as they allow to monitor improvements and regressions introduced by subsequent APIs.
    - So far only `xtimer` and `ztimer` are in the race. Or is there any alternative proposal I'm unaware with an implementation that either is ready for review and testing or will get to this point reasonably soon?

> [...] nor well justified.

Maybe you could point out the requirements which need justification? E.g. the low power requirement should be trivial. Also all functional requirements needed to replace `xtimer` within the RIOT code base should also need no further justification.

tcschmidt commented:
> 
> 
> > I suggest:
> > 
> > * Let's cut out the design decisions section. It can be part of another RDM proposing a specific implementation, or part of the module level documentation we'll end up with.
> 
> Agreed. On a second thought, I think it might not be the brightest idea to document design decisions via an RDM. 

The objective of an RDM is not to document code, but to develop and discuss design ideas and its rationales (see #13943). This seems of  particular importance in cases of repeated design failures such as the timers. 


> Maybe you could point out the requirements which need justification? E.g. the low power requirement should be trivial. Also all functional requirements needed to replace `xtimer` within the RIOT code base should also need no further justification.

I cannot follow any of the _O(n)_ real-time requirements.

maribu reviewed (commented):
Some updates

maribu commented:
> I cannot follow any of the _O(n)_ real-time requirements.

I added a justification for that.

maribu commented:
> This seems of particular importance in cases of repeated design failures such as the timers.

Can you elaborate on why you consider them as design failures?

E.g. when you judge the design of medieval horse carriage by current mobility requirements, those would certainly classify as design failures. Similar cars from the 1960s would be design failures with the current mobility requirements. And likely the best current cars will not be sufficient in terms of the mobility requirements in 2080. Yet, no car manufacturer will try to meet the expectations of people in 2080 as of now.

One could view this as a history of constant design failures... Or as constant progress.

I tend to believe that "RIOT's history of timer API design failures" also is better described as constant progress.

aabadie commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,257 @@
+5. Only available when using TOSTThreads
+6. `xtimer_periodic_wakeup()`/`ztimer_periodic_wakeup()` can be used, but
+   requires a dedicated thread allocated for the periodic task
+7. Mynwest uses a `struct` with a members `tv_sec` (64 bit) and `tv_usec`
```
```suggestion
7. Mynewt uses a `struct` with a members `tv_sec` (64 bit) and `tv_usec`
```

aabadie commented:
> The major contribution of this doc seems the overview of the different timer subsystems in IoT OSs. So a suggestion for a new title reads A Survey of Selected Aspects of Timer Subsystems in IoT OSs.

While I agree that the RDM doesn't contain a precise API description. I disagree with your remark and title proposal: this RDM is not only an overview of the different timer subsystems, it also contains a long list of requirements (features, real-time, precision, performance) for such high-level timers and that an API **in RIOT** should offer.

So compared to the actual RDM title, replacing *API* with *API requirements* would be inline with the content and probably enough: *RIOT's High Level Timer API requirements*.

maribu commented:
I updated the title to contain both the survey and the requirements analysis and also added every contributor as author (sorted by last name, alphabetically).

emmanuelsearch commented:
I agree with @aabadie here. From my point of view, the most crucial part of this document for our next steps is the requirements we set for our high-level timer API. Would be bizarre not to put this in the title of the RDM.

tcschmidt commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,259 @@
+while not being to strict to rule out reasonable trade-offs e.g. between
+ROM/RAM requirements and latency. An *O(n)* latency is e.g. feasible to
+implement with a sorted linked list and therefore a reasonable lower bar.
+
```
*O(n)* is an unbound function, so this justification is inconsistent. For real-time requirements ask for *O(1)*, I suppose.

maribu replied:
For every concrete application, there is an upper bound for the number *n* of software timers being active at the same time. So for every concrete application, *O(n)* has an upper bound.

tcschmidt replied:
This is true for any finite set of numbers. So we could also go for an *O(exp(n))*?

kaspar030 replied:
> _O(n)_ is an unbound function, so this justification is inconsistent. For real-time requirements ask for _O(1)_, I suppose.

This is intentionally written to _not_ be O(1), because of the reasonable trade-offs as described. We know how to do O(n) by simply using linked lists and sorting on insert, and we understand the runtime implications well. We don't know how to do this in O(1) without limiting the number of allowed timers.

We do multiplexing here, so one hardware timer (or more, but a small number) needs to be used to schedule N timeouts.
So when adding one of these timeouts (using a set(interval) function), that new timeout needs to be added into the list of already queued timers. Using linked lists, this is an O(n) operation. Using an array and binary search, this is O(logn). We need to be careful what to require here, because if we break O(n*logn) for sorting n timers, we might attract a lot of attention.

It is possible to add an allowed timers limit (that is lower than available memory, e.g., 16 or 128) and enforce this.
We'd suddenly have O(1) on paper (meeting the requirements), but wouldn't have gained anything, other than that now a multitude of functions have an error case (timer_sleep() might fail with "maximum number of timers reached") that needs to be checked everywhere, which is unreasonable.

TL;DR let's not require something we don't know how to implement. This is phrased as "yadda yadda reasonable trade-offs".

tcschmidt replied:
The complete timer management can be done in O(log(n)), which is almost constant in the number range we consider realistically. Further optimizations can apply, if algorithms account for the deadline of the next firing timer, which can be determined in constant time. 
Current preliminary HiL-tests of @pokgak show a rather poor behavior of xtimer and a somewhat unclear signature of ztimer. We will publish once ready. 
In any case, it is important that timer firing remains within predictable bounds independent of the number of instantiated timers - otherwise real-time gets out of control.

maribu replied:
> The complete timer management can be done in O(log(n)), which is almost constant in the number range we consider realistically.

I thought we all agree on the fact that *O(f(n))* is constant for any *f(n)* within the number range we consider realistically?

tcschmidt replied:
Off course, we don't.

bergzand replied:
> _O(n)_ is an unbound function, so this justification is inconsistent. For real-time requirements ask for _O(1)_, I suppose.

Nope, soft and hard real-time system only care if the growth is bounded at some point, they require a *the worst case* execution time (in units of time or instructions).

The growth function doesn't provide any information on how long the function takes to execute. An _O(1)_ function is just as bad as an n-limited O(n) function if it exceeds the permitted execution time, the only given is that the O(1) will always exceed the deadline and that the O(n) might exceed the deadline in some cases.

As long as the developer can have a guarantee that the number of simultaneous active timers is bounded, the worst case execution time can be determined and whether the system will function can be reasoned about. Doesn't matter here if the growth function is O(1), O(n), O(log(n)) or O(exp(n)) as those do not tell us anything about the worst case execution time, only how it scales if the number of active timers is increased.

tcschmidt replied:
> Doesn't matter here if the growth function is O(1), O(n), O(log(n)) or O(exp(n)) as those do not tell us anything about the worst case execution time, only how it scales if the number of active timers is increased.

???

bergzand replied:
> ???

Care to elaborate what exactly is not clear here?

tcschmidt replied:
log(25) ~ 1.39,  exp(25) ~ 7.2 x 10**10.

bergzand replied:
> log(25) ~ 1.39, exp(25) ~ 7.2 x 10**10.

Both bounded, which is exactly what a realtime system needs.

tcschmidt replied:
Really?

bergzand replied:
> Really?

Do you want to dive into the numerical analysis of the numbers you provided or can we agree here that both numbers are constants?

A realtime system does not require the system to be fast. It does however require the system to respond within the specified deadlines. This deadline can be in the order of microsecond or in the order of years, depending on the requirements.

Furthermore, a complex O(1) algorithm, taking 10 seconds performs worse compared with a simple O(n) algorithm taking `10 ms + n * 1 ms`. when n is known to be bounded to 100.

bergzand replied:
As a solution here, why don't we rephrase a bit and specify that it must be trivial to determine the maximum number of active timers for an application in order to ensure bounded execution time?

kaspar030 commented:
@MichelRottleuthner do you want to add requirements, or are you fine with this for now?
@tcschmidt are you OK with the title now?

jue89 commented on doc/memos/rdm-draft-buschsieweke-timer-api.md:
```diff
@@ -0,0 +1,259 @@
+## Performance Requirements
+
+- Efficient use of RAM/ROM
+- Optimized for power management
```
What do you think of elaborating this a little bit?

As the discussion above already pointed out, it is required to turn off underlying hardware timers if they aren't needed to gain power savings.

I'd add something like this (feel free to polish or rephrase):

```
    - The timer implementation should disable underlying hardware in periods when the timer is not used by the application. (I.e. no software timer is running and the timer doesn't back a system time.)
```

Or does this already touches design decisions too much, and thus is out of scope of this RDM?

maribu replied:
I'd say this is a implementation detail. I fully agree that this is the only sensible way to achieve the goal of proper power management - so only implementations doing just that would IMO be considered for adoption.

jue89 reviewed (changes requested):
I just had a deeper look into this RDM. Its been a while since the last time I did that.

All in all, this RDM covers all features we require for our use cases. A nice piece of work!

Just a minor comment. (Sorry for kicking in this late during the polishing process!)

fjmolinas commented:
ping @maribu!

kaspar030 commented:
This PR started out to be _the_ RDM on timers, so we could base any further work on it with confidence. But currently it is a collection of requirements and some survey on features other timer implementations provide. It does not go into specific design questions for a high-level timer implementation.

At the same time, ztimer got merged and provides many solutions to problems we have with xtimer. But migrating to ztimer as the default timer is a bit stalled for lack of an RDM approved design, which this RDM was supposed to give, but doesn't.

I think we all agree that doing comprehensive research on "how to do high-level timers right[tm]" is a lot of work. Many of us think that ztimer is a step in the right direction, and we should migrate. AFAIK noone thinks xtimer can be turned into whatever we need without substantial rework. And xtimer effectively blocks RIOT from sleeping, which should be a fix first huuuuge bug for an OS supposed to run on low-power devices.

Can we agree on most of the requirements, and concentrate on that for _this_ PR?

We could then move forward with ztimer and check if it meets this PR's requirements, in a follow-up RDM.
Also, work on a proper "how to do high-level timers right[tm]" RDM can start.

(would it make sense to even drop the survey part here and save it for the more extensive research RDM?)

benpicco commented:
> But migrating to ztimer as the default timer is a bit stalled for lack of an RDM approved design, which this RDM was supposed to give, but doesn't.

I don't think the lack of RDM stops ztimer from being used by default.
AFAIR there are still features missing from ztimer (`ztimer_mutex_lock_timeout()`) that prevent it from replacing xtimer entirely.

tcschmidt commented:
Even though I believe a rigorous and exhaustive design document would be essential for reworking the timer API, I don't see any of this here. It rather appears as an - somewhat related - informational document about "some thoughts about timers without rigorous justification". This certainly cannot serve as a foundation for redesign. The question is rather, whether this is helpful and needed or whether this can simply be discarded. 
Any opinions?

maribu commented:
Let me recapitulate the history of this:

1. We have with `xtimer` a high level timer API that is known for years to not be insufficient for any low power scenario
2. With `ztimer` an alternative was proposed that is known to be a huge step in the right direction
3. The transition to `ztimer` was blocked with the argument that an RDM is needed first
4. An RDM was started to address this
5. The goals of the RDM were scaled down to reflect the lack of progress and engagement as well as the inability to reach a consensus on anything more ambitions than a list of basic requirements
    - The small list of features and superficial implementation details of timer APIs of competing OS could also stay (at least so far) - good to know that at least lists unopinionated facts don't spawn endless discussions
6. Even for this unambitious RDM it seems to be difficult to reach a consensus - despite that this list of requirements is relatively unopinionated due to a lack of prioritization of the requirements most likely to end up on opposing sides of trade-offs
7. Just when it seems that finally an agreement on this least ambitions outcome of the RDM is in reach - people start to call for burying the RDM
8. Meanwhile: RIOT remains completely unsuitable for any low power scenarios in default configuration - despite a solid alternative that performs well in low power scenarios is already in the code base

An particularly interesting observation is that the very persons that were the first to call out for an RDM as requirement for any progress in regard to the timer API are seemingly the last persons to contribute to it.

I'm willing to keep pushing this RDM to some sort of conclusion. But please keep me out of the loop for any follow up.

benpicco commented:
> * The transition to `ztimer` was blocked with the argument that an RDM is needed first

Is that really the case? If ztimer works and is a drop in replacement for xtimer, nobody will care about the RDM anymore.
With the 802.15.4 radio HAL the RDM is used to discuss how things should be implemented, then implement them. It's a living document that aids development.
Having an RDM for the sake of having an RDM is pointless.

What *really* blocks transition to `ztimer` is that not all features of xtimer are covered yet.
Once it's a drop in replacement we can just replace `xtimer` with `ztimer_xtimer_compat`.

maribu commented:
> > * The transition to `ztimer` was blocked with the argument that an RDM is needed first
> 
> Is that really the case?

From the discussion in the [devel mailing list](https://lists.riot-os.org/pipermail/devel/2019-December/thread.html) I pretty much got the impression. Here are some quotes of the key arguments leading to this RDM:



> Side note: I think we should move the discussion to an RDM PR for the 
> design document and discuss there to not get lost.

@MichelRottleuthner, 10. 12. 2019, [full email for context](https://lists.riot-os.org/pipermail/devel/2019-December/006404.html)

> What do you think about an RDM PR?

@MichelRottleuthner, 11. 12. 2019, [full email for context](https://lists.riot-os.org/pipermail/devel/2019-December/006406.html)

> I'd favor to not merge it before we worked on the RDM as it may show 
> that there are things that should be done differently.
> 
> [...]
>
> I'll provide the collected information to the RDM.
>
> [...]
>
> > > No, I mean if "having a non functional timer for many years" happens again.
> > > I think the way how xtimer did is job over these years is not something
> > > we want to repeat.  
> >
> > So an RDM would have prevented this? I doubt it. xtimer's ISR buggyness
> > was known *for years* even without an RDM. We were just inexperienced
> > (or ignorant, or incompetent, your pick).  
>
> Yes I honestly think it could have helped to on that end to thoroughly 
> discuss the key problems with a broader audience of interested and 
> competent people.

@MichelRottleuthner, 13. 12. 2019, [full email for context](https://lists.riot-os.org/pipermail/devel/2019-December/006413.html)

@tcschmidt did not explicitly ask for an RDM, but for a "problem statement and design document" and wanted "a clear and falsifiable problem statement" prior to even discussing the adoption of `ztimer`, see [full email](https://lists.riot-os.org/pipermail/devel/2019-December/006393.html).

To my best understanding, both @MichelRottleuthner and @tcschmidt saw completion of an RDM (or in case of @tcschmidt an RDM-like document) as requirement for adoption of `ztimer`. So I started this off:

> in order to push this forward I just opened a PR for an RDM at
> https://github.com/RIOT-OS/RIOT/pull/12970. This PR is in an early state and
> both feedback and help will be greatly appreciated.

@maribu, 16. 12. 2019, [full email for context](https://lists.riot-os.org/pipermail/devel/2019-December/006417.html)

I think it is fair to say that nobody's expectations towards this RDM were met. And honestly, I see no indication that a second try should work out better.

I'm not sure if what @benpicco was suggesting to rather work on improving `ztimer` and fill the gaps rather than endlessly stalling progress because of RDMs that don't get anywhere. But if so: strong +1

Btw: https://github.com/RIOT-OS/RIOT/pull/15052 would be a huge step towards `ztimer_mutex_lock_timeout()`.

waehlisch commented:
> > * The transition to `ztimer` was blocked with the argument that an RDM is needed first
> 
> Is that really the case?

several people asked for an RDM to have a base for an informed discussion. it was not about "Having an RDM for the sake of having an RDM". 

quite often sound solutions need informed discussions, to prevent trial and error. ... as someone, i think it was Kaspar, said in the timer breakout session in Helsinki: why should we be successful this time? we know from the past that code that is not enough.

emmanuelsearch commented:
> I think it is fair to say that nobody's expectations towards this RDM were met. And honestly, I see no indication that a second try should work out better.

@maribu thanks a lot for the work you put in this.

Maybe we are falsely lead to see this glass half empty.
On the positive, half-full side:
- we have collaboratively gathered a comprehensive list of requirements;
- we have a solution designed and partly implemented -- but which already seems to improve the status quo;
- some already use this solution in production.

Let's not lose sight of the above, which can be considered useful progress.
In that respect, I consider this RDM already useful -- independent of its stalling for other reasons.

benpicco commented:
As this seems to be the place to put general ideas about the high-level timer API, here is another one:

I think the high-level timer could be simplified if it configured the low-level timer in down-counting mode.
If we want to set a timeout in `n` ticks, right now we have to

 - get the current time
 - add `n`, make sure we did not overshoot already

instead, with down-counting timers we only have to

- set timeout to `n`

Now this simplifies the `set_timeout()` function at the expense of the `get_time()` function - however, I think the former is executed a lot more often. 

What do you think?

tcschmidt commented:
@benpicco Interesting perspective. I suppose it needs experimental validation, though. 
Also, low-level timers should see more requirements or desires from the high-level side: A lot more interrelation with #15046 would clearly progress insights w.r.t. the real problems of timers.
